django_components/templatetags/component_tags.py

```diff
@@ -160,99 +162,177 @@ def __init__(self, slot: "SlotNode", context: Context):
+                fill_node = implicit_fill_node_stack.pop()
+        if self.is_required:
+            if fill_node is None:
                 raise TemplateSyntaxError(
                     f"Slot '{self.name}' is marked as 'required' (i.e. non-optional), "
                     f"yet no fill is provided. Check template '{self.template_name}'"
                 )
-        else:
-            fill_node = fill_node_stack.pop()
+        # //
+        # Get nodelist and add alias variables to context.
+        extra_context = {}
+        if fill_node:
             nodelist = fill_node.nodelist
-
             if fill_node.alias_var is not None:
                 aliased_slot_var = UserSlotVar(self, context)
                 resolved_alias_name = fill_node.alias_var.resolve(context)
                 extra_context[resolved_alias_name] = aliased_slot_var
+        else:
+            nodelist = self.nodelist  # slot nodelist
+        # //
+
         with context.update(extra_context):
             return nodelist.render(context)
 
 
+SLOT_REQUIRED_OPTION_KEYWORD = "required"
+SLOT_DEFAULT_OPTION_KEYWORD = "default"
+
+
 @register.tag("slot")
 def do_slot(parser, token):
     bits = token.split_contents()
     args = bits[1:]
     # e.g. {% slot <name> %}
-    if len(args) == 1:
-        slot_name: str = args[0]
-        required = False
-    elif len(args) == 2:
-        slot_name: str = args[0]
-        required_keyword = args[1]
-        if required_keyword != "required":
+    is_required = False
+    is_default = False
+    if 1 <= len(args) <= 3:
+        slot_name, *options = args
+        if not is_wrapped_in_quotes(slot_name):
             raise TemplateSyntaxError(
-                f"'{bits[0]}' only accepts 'required' keyword as optional second argument"
+                f"'{bits[0]}' name must be a string 'literal'."
+            )
+        slot_name = strip_quotes(slot_name)
+        raise_if_not_py_identifier(slot_name, bits[0])
+        modifiers_count = len(options)
+        if SLOT_REQUIRED_OPTION_KEYWORD in options:
+            is_required = True
+            modifiers_count -= 1
+        if SLOT_DEFAULT_OPTION_KEYWORD in options:
+            is_default = True
+            modifiers_count -= 1
+        if modifiers_count != 0:
+            keywords = [
+                SLOT_REQUIRED_OPTION_KEYWORD,
+                SLOT_DEFAULT_OPTION_KEYWORD,
+            ]
+            raise TemplateSyntaxError(
+                f"Invalid options passed to 'slot' tag. Valid choices: {keywords}."
             )
-        else:
-            required = True
     else:
         raise TemplateSyntaxError(
-            f"{bits[0]}' tag takes only one argument (the slot name)"
-        )
-
-    if not is_wrapped_in_quotes(slot_name):
-        raise TemplateSyntaxError(
-            f"'{bits[0]}' name must be a string 'literal'."
+            "'slot' tag does not match pattern {{% slot <name> ['default'] ['required'] %}}. "
+            "Order of options is free."
```

Comment:
Free is not the right term here I think, maybe remove this line?

Reply:
It's probably too much detail for an error msg anyway. What I mean is that 'default' can come before or after 'required'. In ebnf terms:

```ebnf
slot tag = name, [options];
options = ["default"], ["required"] | ["required"], ["default"];
```

I'll drop it and let the examples speak for themselves.